Build script fails when cross-compiling to `aarch64-unknown-linux-gnu`

Coming from #42:

Hi, I'm having the same issue with targeting aarch64-unknown-linux-gnu. I've gathered that the "original" issue was https://github.com/jean-airoldie/zeromq-src-rs/issues/28 about strlcpy-detection in the original C-library's cmake build system vs. here. I've also followed https://github.com/erickt/rust-zmq/issues/385 to [this commit](https://github.com/jean-airoldie/zeromq-src-rs/commit/21056e972a91b96bc98969fe76626585695410ed) and understood that the fix that was made there to support cross-compilation requires using a version of zeromq-src >= 0.3, which is not yet used by a published zmq release.

However, I then tried to locally patch the `zmq-sys` dependency to point to the branch from https://github.com/erickt/rust-zmq/pull/386 using

```toml
[patch.crates-io]
zmq-sys = { git = "https://github.com/MarijnS95/rust-zmq.git", branch = "zeromq-src-0.3" }
```

This was successful in causing the version of zeromq-rs used during the build to be bumped to zeromq-src-0.3.2+4.3.5 and would compile on the host (as does the older version), but still fails to cross-compile to aarch64 giving

```
error: failed to run custom build command for `zmq-sys v0.12.0 (https://github.com/MarijnS95/rust-zmq.git?branch=zeromq-src-0.3#1aba88ba)`

Caused by:
  process didn't exit successfully: `/target/debug/build/zmq-sys-6dabaad22ef69b95/build-script-main` (exit status: 1)

  --- stdout omitted
  --- stderr
 /cargo/registry/src/index.crates.io-6f17d22bba15001f/zeromq-src-0.3.2+4.3.5/src/strlcpy.c: In function 'main':
  /cargo/registry/src/index.crates.io-6f17d22bba15001f/zeromq-src-0.3.2+4.3.5/src/strlcpy.c:5:11: warning: implicit declaration of function 'strlcpy' [-Wimplicit-function-declaration]
       (void)strlcpy(buf, "a", 1);
             ^
  /tmp/ccWhvypa.o: In function `main':
  /cargo/registry/src/index.crates.io-6f17d22bba15001f/zeromq-src-0.3.2+4.3.5/src/strlcpy.c:5: undefined reference to `strlcpy'
  collect2: error: ld returned 1 exit status


  error occurred: Command "aarch64-linux-gnu-g++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "-I" "/cargo/registry/src/index.crates.io-6f17d22bba15001f/zeromq-src-0.3.2+4.3.5/vendor/include" "-I" "/cargo/registry/src/index.crates.io-6f17d22bba15001f/zeromq-src-0.3.2+4.3.5/vendor/src" "-I" "/cargo/registry/src/index.crates.io-6f17d22bba15001f/zeromq-src-0.3.2+4.3.5/vendor/src" "-I" "/cargo/registry/src/index.crates.io-6f17d22bba15001f/zeromq-src-0.3.2+4.3.5/vendor/external/sha1" "-I" "/target/aarch64-unknown-linux-gnu/debug/build/zmq-sys-a7433891bd83dba9/out" "-Wall" "-Wextra" "-DZMQ_BUILD_TESTS=OFF" "-DZMQ_USE_CV_IMPL_STL11=1" "-DZMQ_STATIC=1" "-DZMQ_USE_BUILTIN_SHA1=1" "-DZMQ_HAVE_WS=1" "-DZMQ_IOTHREAD_POLLER_USE_EPOLL=1" "-DZMQ_POLL_BASED_ON_POLL=1" "-DZMQ_HAVE_IPC=1" "-DHAVE_STRNLEN=1" "-DZMQ_HAVE_UIO=1" "-DZMQ_HAVE_STRUCT_SOCKADDR_UN=1" "-o" "/target/aarch64-unknown-linux-gnu/debug/build/zmq-sys-a7433891bd83dba9/out/lib/2d62dae1cebcdbc5-socket_base.o" "-c" "/cargo/registry/src/index.crates.io-6f17d22bba15001f/zeromq-src-0.3.2+4.3.5/vendor/src/socket_base.cpp" with args "aarch64-linux-gnu-g++" did not execute successfully (status code exit status: 1).
```

The error clearly references main in src/strlcpy.c. I guess (as @jean-airoldie mentions) it is expected that the "test build" of that file would fail on the target if its glibc does not define strlcpy, but it appears that somehow that also causes the entire build to fail.


<details>

<summary>Full compiler output</summary>

```
The following warnings were emitted during compilation:

warning: zmq-sys@0.12.0: In file included from /usr/aarch64-linux-gnu/include/c++/5/condition_variable:35:0,
warning: zmq-sys@0.12.0:                  from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:70,
warning: zmq-sys@0.12.0:                  from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/mailbox_safe.hpp:16,
warning: zmq-sys@0.12.0:                  from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/mailbox_safe.cpp:4:
warning: zmq-sys@0.12.0: /usr/aarch64-linux-gnu/include/c++/5/bits/c++0x_warning.h:32:2: error: #error This file requires compiler and library support for the ISO C++ 2011 standard. This support must be enabled with the -std=c++11 or -std=gnu++11 compiler options.
warning: zmq-sys@0.12.0:  #error This file requires compiler and library support \
warning: zmq-sys@0.12.0:   ^
warning: zmq-sys@0.12.0: In file included from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/mailbox_safe.hpp:16:0,
warning: zmq-sys@0.12.0:                  from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/mailbox_safe.cpp:4:
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:102:10: error: 'condition_variable_any' in namespace 'std' does not name a type
warning: zmq-sys@0.12.0:      std::condition_variable_any _cv;
warning: zmq-sys@0.12.0:           ^
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp: In member function 'int zmq::condition_variable_t::wait(zmq::mutex_t*, int)':
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:84:13: error: '_cv' was not declared in this scope
warning: zmq-sys@0.12.0:              _cv.wait (
warning: zmq-sys@0.12.0:              ^
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:86:20: error: '_cv' was not declared in this scope
warning: zmq-sys@0.12.0:          } else if (_cv.wait_for (*mutex_, std::chrono::milliseconds (timeout_))
warning: zmq-sys@0.12.0:                     ^
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:86:48: error: 'std::chrono' has not been declared
warning: zmq-sys@0.12.0:          } else if (_cv.wait_for (*mutex_, std::chrono::milliseconds (timeout_))
warning: zmq-sys@0.12.0:                                                 ^
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:87:28: error: 'std::cv_status' has not been declared
warning: zmq-sys@0.12.0:                     == std::cv_status::timeout) {
warning: zmq-sys@0.12.0:                             ^
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp: In member function 'void zmq::condition_variable_t::broadcast()':
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:98:9: error: '_cv' was not declared in this scope
warning: zmq-sys@0.12.0:          _cv.notify_all ();
warning: zmq-sys@0.12.0:          ^
warning: zmq-sys@0.12.0: ToolExecError: Command "aarch64-linux-gnu-g++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/include" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/external/sha1" "-I" "/target/aarch64-unknown-linux-gnu/debug/build/zmq-sys-8d8854777a617d31/out" "-Wall" "-Wextra" "-DZMQ_BUILD_TESTS=OFF" "-DZMQ_USE_CV_IMPL_STL11=1" "-DZMQ_STATIC=1" "-DZMQ_USE_BUILTIN_SHA1=1" "-DZMQ_HAVE_WS=1" "-DZMQ_IOTHREAD_POLLER_USE_EPOLL=1" "-DZMQ_POLL_BASED_ON_POLL=1" "-DZMQ_HAVE_IPC=1" "-DHAVE_STRNLEN=1" "-DZMQ_HAVE_UIO=1" "-DZMQ_HAVE_STRUCT_SOCKADDR_UN=1" "-o" "/target/aarch64-unknown-linux-gnu/debug/build/zmq-sys-8d8854777a617d31/out/lib/b5831782cc127234-mailbox_safe.o" "-c" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/mailbox_safe.cpp" with args aarch64-linux-gnu-g++ did not execute successfully (status code exit status: 1).
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/proxy.cpp: In function 'int zmq::proxy_steerable(zmq::socket_base_t*, zmq::socket_base_t*, zmq::socket_base_t*, zmq::socket_base_t*)':
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/proxy.cpp:502:27: warning: missing initializer for member 'stats_socket::bytes' [-Wmissing-field-initializers]
warning: zmq-sys@0.12.0:      stats_proxy stats = {0};
warning: zmq-sys@0.12.0:                            ^
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/proxy.cpp:502:27: warning: missing initializer for member 'stats_endpoint::recv' [-Wmissing-field-initializers]
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/proxy.cpp:502:27: warning: missing initializer for member 'stats_proxy::backend' [-Wmissing-field-initializers]
warning: zmq-sys@0.12.0: In file included from /usr/aarch64-linux-gnu/include/c++/5/condition_variable:35:0,
warning: zmq-sys@0.12.0:                  from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:70,
warning: zmq-sys@0.12.0:                  from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/mailbox_safe.hpp:16,
warning: zmq-sys@0.12.0:                  from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/socket_base.cpp:47:
warning: zmq-sys@0.12.0: /usr/aarch64-linux-gnu/include/c++/5/bits/c++0x_warning.h:32:2: error: #error This file requires compiler and library support for the ISO C++ 2011 standard. This support must be enabled with the -std=c++11 or -std=gnu++11 compiler options.
warning: zmq-sys@0.12.0:  #error This file requires compiler and library support \
warning: zmq-sys@0.12.0:   ^
warning: zmq-sys@0.12.0: In file included from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/mailbox_safe.hpp:16:0,
warning: zmq-sys@0.12.0:                  from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/socket_base.cpp:47:
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:102:10: error: 'condition_variable_any' in namespace 'std' does not name a type
warning: zmq-sys@0.12.0:      std::condition_variable_any _cv;
warning: zmq-sys@0.12.0:           ^
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp: In member function 'int zmq::condition_variable_t::wait(zmq::mutex_t*, int)':
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:84:13: error: '_cv' was not declared in this scope
warning: zmq-sys@0.12.0:              _cv.wait (
warning: zmq-sys@0.12.0:              ^
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:86:20: error: '_cv' was not declared in this scope
warning: zmq-sys@0.12.0:          } else if (_cv.wait_for (*mutex_, std::chrono::milliseconds (timeout_))
warning: zmq-sys@0.12.0:                     ^
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:86:48: error: 'std::chrono' has not been declared
warning: zmq-sys@0.12.0:          } else if (_cv.wait_for (*mutex_, std::chrono::milliseconds (timeout_))
warning: zmq-sys@0.12.0:                                                 ^
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:87:28: error: 'std::cv_status' has not been declared
warning: zmq-sys@0.12.0:                     == std::cv_status::timeout) {
warning: zmq-sys@0.12.0:                             ^
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp: In member function 'void zmq::condition_variable_t::broadcast()':
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:98:9: error: '_cv' was not declared in this scope
warning: zmq-sys@0.12.0:          _cv.notify_all ();
warning: zmq-sys@0.12.0:          ^
warning: zmq-sys@0.12.0: ToolExecError: Command "aarch64-linux-gnu-g++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/include" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/external/sha1" "-I" "/target/aarch64-unknown-linux-gnu/debug/build/zmq-sys-8d8854777a617d31/out" "-Wall" "-Wextra" "-DZMQ_BUILD_TESTS=OFF" "-DZMQ_USE_CV_IMPL_STL11=1" "-DZMQ_STATIC=1" "-DZMQ_USE_BUILTIN_SHA1=1" "-DZMQ_HAVE_WS=1" "-DZMQ_IOTHREAD_POLLER_USE_EPOLL=1" "-DZMQ_POLL_BASED_ON_POLL=1" "-DZMQ_HAVE_IPC=1" "-DHAVE_STRNLEN=1" "-DZMQ_HAVE_UIO=1" "-DZMQ_HAVE_STRUCT_SOCKADDR_UN=1" "-o" "/target/aarch64-unknown-linux-gnu/debug/build/zmq-sys-8d8854777a617d31/out/lib/b5831782cc127234-socket_base.o" "-c" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/socket_base.cpp" with args aarch64-linux-gnu-g++ did not execute successfully (status code exit status: 1).
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/zmq.cpp:1355:32: warning: unused parameter 'items_' [-Wunused-parameter]
warning: zmq-sys@0.12.0:  int zmq_ppoll (zmq_pollitem_t *items_,
warning: zmq-sys@0.12.0:                                 ^
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/zmq.cpp:1356:20: warning: unused parameter 'nitems_' [-Wunused-parameter]
warning: zmq-sys@0.12.0:                 int nitems_,
warning: zmq-sys@0.12.0:                     ^
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/zmq.cpp:1357:21: warning: unused parameter 'timeout_' [-Wunused-parameter]
warning: zmq-sys@0.12.0:                 long timeout_,
warning: zmq-sys@0.12.0:                      ^
warning: zmq-sys@0.12.0: /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/zmq.cpp:1358:32: warning: unused parameter 'sigmask_' [-Wunused-parameter]
warning: zmq-sys@0.12.0:                 const sigset_t *sigmask_)
warning: zmq-sys@0.12.0:                                 ^

error: failed to run custom build command for `zmq-sys v0.12.0 (/home/<user>/dev/rust-zmq/zmq-sys)`

Caused by:
  process didn't exit successfully: `/target/debug/build/zmq-sys-dd487563397fb08a/build-script-main` (exit status: 1)
  --- stdout
  cargo:rerun-if-changed=build/main.rs
  cargo:rerun-if-env-changed=PROFILE
  cargo:rerun-if-changed=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/src/strlcpy.c
  TARGET = Some(aarch64-unknown-linux-gnu)
  OPT_LEVEL = Some(0)
  HOST = Some(x86_64-unknown-linux-gnu)
  cargo:rerun-if-env-changed=CXX_aarch64-unknown-linux-gnu
  CXX_aarch64-unknown-linux-gnu = None
  cargo:rerun-if-env-changed=CXX_aarch64_unknown_linux_gnu
  CXX_aarch64_unknown_linux_gnu = Some(aarch64-linux-gnu-g++)
  RUSTC_WRAPPER = None
  cargo:rerun-if-env-changed=CC_ENABLE_DEBUG_OUTPUT
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some(true)
  CARGO_CFG_TARGET_FEATURE = Some(neon)
  cargo:rerun-if-env-changed=CXXFLAGS_aarch64-unknown-linux-gnu
  CXXFLAGS_aarch64-unknown-linux-gnu = None
  cargo:rerun-if-env-changed=CXXFLAGS_aarch64_unknown_linux_gnu
  CXXFLAGS_aarch64_unknown_linux_gnu = None
  cargo:rerun-if-env-changed=TARGET_CXXFLAGS
  TARGET_CXXFLAGS = None
  cargo:rerun-if-env-changed=CXXFLAGS
  CXXFLAGS = None
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  cargo:warning=In file included from /usr/aarch64-linux-gnu/include/c++/5/condition_variable:35:0,
  cargo:warning=                 from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:70,
  cargo:warning=                 from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/mailbox_safe.hpp:16,
  cargo:warning=                 from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/mailbox_safe.cpp:4:
  cargo:warning=/usr/aarch64-linux-gnu/include/c++/5/bits/c++0x_warning.h:32:2: error: #error This file requires compiler and library support for the ISO C++ 2011 standard. This support must be enabled with the -std=c++11 or -std=gnu++11 compiler options.
  cargo:warning= #error This file requires compiler and library support \
  cargo:warning=  ^
  cargo:warning=In file included from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/mailbox_safe.hpp:16:0,
  cargo:warning=                 from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/mailbox_safe.cpp:4:
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:102:10: error: 'condition_variable_any' in namespace 'std' does not name a type
  cargo:warning=     std::condition_variable_any _cv;
  cargo:warning=          ^
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp: In member function 'int zmq::condition_variable_t::wait(zmq::mutex_t*, int)':
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:84:13: error: '_cv' was not declared in this scope
  cargo:warning=             _cv.wait (
  cargo:warning=             ^
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:86:20: error: '_cv' was not declared in this scope
  cargo:warning=         } else if (_cv.wait_for (*mutex_, std::chrono::milliseconds (timeout_))
  cargo:warning=                    ^
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:86:48: error: 'std::chrono' has not been declared
  cargo:warning=         } else if (_cv.wait_for (*mutex_, std::chrono::milliseconds (timeout_))
  cargo:warning=                                                ^
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:87:28: error: 'std::cv_status' has not been declared
  cargo:warning=                    == std::cv_status::timeout) {
  cargo:warning=                            ^
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp: In member function 'void zmq::condition_variable_t::broadcast()':
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:98:9: error: '_cv' was not declared in this scope
  cargo:warning=         _cv.notify_all ();
  cargo:warning=         ^
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 1
  cargo:warning=ToolExecError: Command "aarch64-linux-gnu-g++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/include" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/external/sha1" "-I" "/target/aarch64-unknown-linux-gnu/debug/build/zmq-sys-8d8854777a617d31/out" "-Wall" "-Wextra" "-DZMQ_BUILD_TESTS=OFF" "-DZMQ_USE_CV_IMPL_STL11=1" "-DZMQ_STATIC=1" "-DZMQ_USE_BUILTIN_SHA1=1" "-DZMQ_HAVE_WS=1" "-DZMQ_IOTHREAD_POLLER_USE_EPOLL=1" "-DZMQ_POLL_BASED_ON_POLL=1" "-DZMQ_HAVE_IPC=1" "-DHAVE_STRNLEN=1" "-DZMQ_HAVE_UIO=1" "-DZMQ_HAVE_STRUCT_SOCKADDR_UN=1" "-o" "/target/aarch64-unknown-linux-gnu/debug/build/zmq-sys-8d8854777a617d31/out/lib/b5831782cc127234-mailbox_safe.o" "-c" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/mailbox_safe.cpp" with args aarch64-linux-gnu-g++ did not execute successfully (status code exit status: 1).
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/proxy.cpp: In function 'int zmq::proxy_steerable(zmq::socket_base_t*, zmq::socket_base_t*, zmq::socket_base_t*, zmq::socket_base_t*)':
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/proxy.cpp:502:27: warning: missing initializer for member 'stats_socket::bytes' [-Wmissing-field-initializers]
  cargo:warning=     stats_proxy stats = {0};
  cargo:warning=                           ^
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/proxy.cpp:502:27: warning: missing initializer for member 'stats_endpoint::recv' [-Wmissing-field-initializers]
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/proxy.cpp:502:27: warning: missing initializer for member 'stats_proxy::backend' [-Wmissing-field-initializers]
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  cargo:warning=In file included from /usr/aarch64-linux-gnu/include/c++/5/condition_variable:35:0,
  cargo:warning=                 from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:70,
  cargo:warning=                 from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/mailbox_safe.hpp:16,
  cargo:warning=                 from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/socket_base.cpp:47:
  cargo:warning=/usr/aarch64-linux-gnu/include/c++/5/bits/c++0x_warning.h:32:2: error: #error This file requires compiler and library support for the ISO C++ 2011 standard. This support must be enabled with the -std=c++11 or -std=gnu++11 compiler options.
  cargo:warning= #error This file requires compiler and library support \
  cargo:warning=  ^
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  cargo:warning=In file included from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/mailbox_safe.hpp:16:0,
  cargo:warning=                 from /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/socket_base.cpp:47:
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:102:10: error: 'condition_variable_any' in namespace 'std' does not name a type
  cargo:warning=     std::condition_variable_any _cv;
  cargo:warning=          ^
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp: In member function 'int zmq::condition_variable_t::wait(zmq::mutex_t*, int)':
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:84:13: error: '_cv' was not declared in this scope
  cargo:warning=             _cv.wait (
  cargo:warning=             ^
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:86:20: error: '_cv' was not declared in this scope
  cargo:warning=         } else if (_cv.wait_for (*mutex_, std::chrono::milliseconds (timeout_))
  cargo:warning=                    ^
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:86:48: error: 'std::chrono' has not been declared
  cargo:warning=         } else if (_cv.wait_for (*mutex_, std::chrono::milliseconds (timeout_))
  cargo:warning=                                                ^
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:87:28: error: 'std::cv_status' has not been declared
  cargo:warning=                    == std::cv_status::timeout) {
  cargo:warning=                            ^
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp: In member function 'void zmq::condition_variable_t::broadcast()':
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/condition_variable.hpp:98:9: error: '_cv' was not declared in this scope
  cargo:warning=         _cv.notify_all ();
  cargo:warning=         ^
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 1
  cargo:warning=ToolExecError: Command "aarch64-linux-gnu-g++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/include" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/external/sha1" "-I" "/target/aarch64-unknown-linux-gnu/debug/build/zmq-sys-8d8854777a617d31/out" "-Wall" "-Wextra" "-DZMQ_BUILD_TESTS=OFF" "-DZMQ_USE_CV_IMPL_STL11=1" "-DZMQ_STATIC=1" "-DZMQ_USE_BUILTIN_SHA1=1" "-DZMQ_HAVE_WS=1" "-DZMQ_IOTHREAD_POLLER_USE_EPOLL=1" "-DZMQ_POLL_BASED_ON_POLL=1" "-DZMQ_HAVE_IPC=1" "-DHAVE_STRNLEN=1" "-DZMQ_HAVE_UIO=1" "-DZMQ_HAVE_STRUCT_SOCKADDR_UN=1" "-o" "/target/aarch64-unknown-linux-gnu/debug/build/zmq-sys-8d8854777a617d31/out/lib/b5831782cc127234-socket_base.o" "-c" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/socket_base.cpp" with args aarch64-linux-gnu-g++ did not execute successfully (status code exit status: 1).
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/zmq.cpp:1355:32: warning: unused parameter 'items_' [-Wunused-parameter]
  cargo:warning= int zmq_ppoll (zmq_pollitem_t *items_,
  cargo:warning=                                ^
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/zmq.cpp:1356:20: warning: unused parameter 'nitems_' [-Wunused-parameter]
  cargo:warning=                int nitems_,
  cargo:warning=                    ^
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/zmq.cpp:1357:21: warning: unused parameter 'timeout_' [-Wunused-parameter]
  cargo:warning=                long timeout_,
  cargo:warning=                     ^
  cargo:warning=/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/zmq.cpp:1358:32: warning: unused parameter 'sigmask_' [-Wunused-parameter]
  cargo:warning=                const sigset_t *sigmask_)
  cargo:warning=                                ^
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0

  --- stderr
  /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/src/strlcpy.c: In function 'main':
  /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/src/strlcpy.c:5:11: warning: implicit declaration of function 'strlcpy' [-Wimplicit-function-declaration]
       (void)strlcpy(buf, "a", 1);
             ^
  /tmp/ccL4fWFZ.o: In function `main':
  /cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/src/strlcpy.c:5: undefined reference to `strlcpy'
  collect2: error: ld returned 1 exit status


  error occurred: Command "aarch64-linux-gnu-g++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/include" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src" "-I" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/external/sha1" "-I" "/target/aarch64-unknown-linux-gnu/debug/build/zmq-sys-8d8854777a617d31/out" "-Wall" "-Wextra" "-DZMQ_BUILD_TESTS=OFF" "-DZMQ_USE_CV_IMPL_STL11=1" "-DZMQ_STATIC=1" "-DZMQ_USE_BUILTIN_SHA1=1" "-DZMQ_HAVE_WS=1" "-DZMQ_IOTHREAD_POLLER_USE_EPOLL=1" "-DZMQ_POLL_BASED_ON_POLL=1" "-DZMQ_HAVE_IPC=1" "-DHAVE_STRNLEN=1" "-DZMQ_HAVE_UIO=1" "-DZMQ_HAVE_STRUCT_SOCKADDR_UN=1" "-o" "/target/aarch64-unknown-linux-gnu/debug/build/zmq-sys-8d8854777a617d31/out/lib/b5831782cc127234-socket_base.o" "-c" "/cargo/git/checkouts/zeromq-src-rs-913d92adcfd7e1e1/b886af7/vendor/src/socket_base.cpp" with args aarch64-linux-gnu-g++ did not execute successfully (status code exit status: 1).


warning: build failed, waiting for other jobs to finish...
```

</details>

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -100,7 +100,8 @@ mod cxx11 {
             PathBuf::from(env::var("OUT_DIR").unwrap()).join("has_cxx11");
 
         cc::Build::new()
-            .warnings(false)
+            .warnings(true)
+            .warnings_into_errors(true)
             .std("c++11")
             .get_compiler()
             .to_command()

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -84,6 +84,35 @@ mod glibc {
     }
 }
 
+mod cxx11 {
+    use std::{
+        env,
+        path::{Path, PathBuf},
+    };
+
+    // Attempt to compile a c program that links has the c++11 flag to determine
+    // whether it is supported.
+    pub(crate) fn has_cxx11() -> bool {
+        let src = Path::new(env!("CARGO_MANIFEST_DIR")).join("src/trivial.c");
+        println!("cargo:rerun-if-changed={}", src.display());
+
+        let dest =
+            PathBuf::from(env::var("OUT_DIR").unwrap()).join("has_cxx11");
+
+        cc::Build::new()
+            .warnings(false)
+            .std("c++11")
+            .get_compiler()
+            .to_command()
+            .arg(src)
+            .arg("-o")
+            .arg(dest)
+            .status()
+            .expect("failed to execute gcc")
+            .success()
+    }
+}
+
 /// The location of a library.
 #[derive(Debug, Clone)]
 pub struct LibLocation {
@@ -441,6 +470,16 @@ impl Build {
         if has_strlcpy {
             build.define("ZMQ_HAVE_STRLCPY", "1");
         }
+
+        // MSVC does not support c++11, since c++14 is the minimum.
+        if !target.contains("msvc") {
+            // Enable c++11 if possible to fix issue 45
+            // (https://github.com/jean-airoldie/zeromq-src-rs/issues/45).
+            if cxx11::has_cxx11() {
+                build.std("c++11");
+            }
+        }
+
         let out_dir = PathBuf::from(env::var("OUT_DIR").unwrap());
         let lib_dir = out_dir.join("lib");
 

--- a/src/trivial.c
+++ b/src/trivial.c
@@ -1,0 +1,3 @@
+int main(void) {
+    return 0;
+}

